lib/ddtrace/configuration/components.rb

```diff
@@ -54,28 +54,116 @@ def build_tracer(settings, agent_settings)
           tracer = settings.tracer.instance
           return tracer unless tracer.nil?
 
-          tracer = Tracer.new(
+          # Apply test mode settings if test mode is activated
+          if settings.test_mode.enabled
+            context_flush = build_test_mode_context_flush(settings)
+            sampler = build_test_mode_sampler
+            writer = build_test_mode_writer(settings, agent_settings)
+          else
+            context_flush = build_context_flush(settings)
+            sampler = build_sampler(settings)
+            writer = build_writer(settings, agent_settings)
+          end
+
+          subscribe_to_writer_events!(writer, sampler)
+
+          Tracer.new(
             default_service: settings.service,
             enabled: settings.tracer.enabled,
-            partial_flush: settings.tracer.partial_flush.enabled,
+            context_flush: context_flush,
+            sampler: sampler,
+            writer: writer,
             tags: build_tracer_tags(settings),
-            sampler: PrioritySampler.new(
+          )
+        end
+
+        def build_context_flush(settings)
+          if settings.tracer.partial_flush.enabled
+            Datadog::ContextFlush::Partial.new(min_spans_before_partial_flush: settings.tracer.partial_flush.min_spans_threshold)
+          else
+            Datadog::ContextFlush::Finished.new
+          end
+        end
+
+        # TODO: Sampler should be a top-level component.
+        # It is currently part of the Tracer initialization
+        # process, but can take a variety of options (including
+        # a fully custom instance) that makes the Tracer
+        # initialization process complex.
+        def build_sampler(settings)
```

Comment:
I'm curious about this comment -- what would be the use of an independent sampler that exists separately from a tracer? (Same question for the writer)

Reply:
Yeah, something like that.
The sampler object has so many complex options (it's basically a rule engine), that it made the tracer configuration too heavy.

Reply:
Hmm, I guess I see what you mean in terms of complexity, but I guess part of my confusion comes from what _being top-level_ means; in particular, and looking at the code below:

```ruby
      attr_reader \
        :health_metrics,
        :logger,
        :profiler,
        :runtime_metrics,
        :tracer

      def initialize(settings)
        # Logger
        @logger = self.class.build_logger(settings)

        agent_settings = AgentSettingsResolver.call(settings, logger: @logger)

        # Tracer
        @tracer = self.class.build_tracer(settings, agent_settings)

        # Profiler
        @profiler = self.class.build_profiler(settings, agent_settings, @tracer)

        # Runtime metrics
        @runtime_metrics = self.class.build_runtime_metrics_worker(settings)

        # Health metrics
        @health_metrics = self.class.build_health_metrics(settings)
      end
```

...when you mean "top-level", do you mean having `@sampler = ...` and `@writer = ...` here as well?

Because it doesn't seem quite correct to me, since while they are quite big/complex, they still seem like implementation details of tracing...?

Reply:
Because they can be completely overwritten by the user with custom instances, it makes sense for them to be easy to track, and not overly attached to our tracing internals.
They are still tracer specific.
Maybe they could live under `Components::Tracing#writer` instead of `Components#writer`.

Reply:
> Because they can be completely overwritten by the user with custom instances, it makes sense for them to be easy to track, and not overly attached to our tracing internals.

Ah. This is a pretty good reason. I somewhat question _that_ decision, but I feel that ship has sailed, so given that constraint, I see the point of giving them a more important standing.